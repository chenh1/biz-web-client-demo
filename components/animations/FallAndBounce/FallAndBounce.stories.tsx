import type { FallAndBounceType } from './FallAndBounce';
import React from 'react';
import { FallAndBounce } from './FallAndBounce';
import { Box } from '../../containers/Box/Box';
import { Heading } from '../../core/Heading/Heading';
import { Paragraph } from '../../core/Paragraph/Paragraph';
import { Button } from '../../core/Button/Button';

type Args = {
  args: FallAndBounceType
}

const Template = (args: FallAndBounceType) => (
  <FallAndBounce height="300px" width="500px" className={"max-w-xs"} {...args}>
    <Box className={"max-w-sm"} hasShadow isRounded gap="sm">
      <Heading type="h3">Heading text</Heading>
      <Paragraph>Example description</Paragraph>
      <Button label="CLICK ME" primary/>
    </Box>
  </FallAndBounce>
)

export default {
  component: Template
}

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const FallAndBounceAnimationFromLeft: Args = { args: { disable: false, from: "fromLeft" } };
export const FallAndBounceAnimationFromRight: Args = { args: { disable: false, from: "fromRight" } };
export const FallAndBounceAnimationFromTop: Args = { args: { disable: false, from: "fromTop" } };
export const FallAndBounceAnimationFromBottom: Args = { args: { disable: false, from: "fromBottom" } };


