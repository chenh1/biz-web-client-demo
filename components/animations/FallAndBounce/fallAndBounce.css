@keyframes fallFromTop {
  0% {
    bottom: 100%;
  }
  60% {
    bottom: 0%;
    transform: scale(1, 1);
  }
  61% {
    transform: scale(1.08, .92);
  }
  75% {
    transform: scale(.95, 1.05);
  }
  88% {
    transform: scale(1.03, .97);
  }
  100% {
    transform: scale(1, 1);
  }
}

@keyframes riseFromBottom {
  0% {
    top: 100%;
  }
  60% {
    top: 0%;
    transform: scale(1, 1);
  }
  61% {
    transform: scale(1.08, .92);
  }
  75% {
    transform: scale(.95, 1.05);
  }
  88% {
    transform: scale(1.03, .97);
  }
  100% {
    transform: scale(1, 1);
  }
}

@keyframes slideFromRight {
  0% {
    left: 100%;
  }
  60% {
    left: 0%;
    transform: scale(1, 1);
  }
  61% {
    transform: scale(.92, 1.08);
  }
  75% {
    transform: scale(1.05, .95);
  }
  88% {
    transform: scale(.97, 1.03);
  }
  100% {
    transform: scale(1, 1);
  }
}

@keyframes slideFromLeft {
  0% {
    right: 100%;
  }
  60% {
    right: 0%;
    transform: scale(1, 1);
  }
  61% {
    transform: scale(.92, 1.08);
  }
  75% {
    transform: scale(1.05, .95);
  }
  88% {
    transform: scale(.97, 1.03);
  }
  100% {
    transform: scale(1, 1);
  }
}



.fallAndBounce > * {
  position: absolute;
}

.fallAndBounce.fromTop > * {
  transform-origin: bottom center;
  animation: fallFromTop 1s ease-in;
  bottom: 0;
}

.fallAndBounce.fromBottom > * {
  transform-origin: top center;
  animation: riseFromBottom 1s ease-in;
  top: 0;
}

.fallAndBounce.fromRight > * {
  transform-origin: center left;
  animation: slideFromRight 1s ease-in;
  left: 0;
}

.fallAndBounce.fromLeft > * {
  transform-origin: center right;
  animation: slideFromLeft 1s ease-in;
  right: 0;
}

.freeze {
  animation-play-state: paused !important;
}