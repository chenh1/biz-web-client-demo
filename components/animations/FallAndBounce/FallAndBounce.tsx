'use client'

import type { AnimationBase, FromDirection } from '../../constants/types';
import React, { useMemo, useState } from 'react';
import { useIntersectionObserver } from '../../hooks/useIntersectionObserver';
import "./fallAndBounce.css"

const getPositionClass = (from: FromDirection): string => {
  switch(from) {
    case 'fromLeft': {
      return 'right-0'
    }
    case 'fromRight': {
      return 'left-0'
    }
    case 'fromTop': {
      return 'bottom-0'
    }
    case 'fromBottom': {
      return 'top-0'
    }
  }
}

export type FallAndBounceType = AnimationBase & {
  from: FromDirection,
  height?: number | `${number}px`,
  width?: number | `${number}px`,
  trackHeight?: number | `${number}px`,
  trackWidth?: number | `${number}px`,
}

export const FallAndBounce: React.FC<FallAndBounceType> = ({ disable, animateWhenInView = true, height = 0, width = 0, trackHeight = 0, trackWidth = 0, className = "", from = "fromLeft", children }) => {
  const computedClasses = useMemo(() => {
    const positionClass = getPositionClass(from);
    return [positionClass].join(' ');
  }, [from]);

  const [ isInView, setIsInView ] = useState(false);

  const moduleRef = useIntersectionObserver({
    inViewCallback: () => {
      setIsInView(true)
    },
    leaveViewCallback: () => {
      setIsInView(false)
    }
  })

  return (
    <div ref={moduleRef} className={`relative ${height ? `h-[${height}]` : ''} ${width ? `w-[${width}]` : ''}`}>
      <div className={`${!disable ? 'fallAndBounce' : ''} ${!animateWhenInView || (animateWhenInView && isInView) ? '' : 'freeze'} absolute ${from} ${computedClasses} ${trackHeight ? `h-[${trackHeight}]` : from === "fromBottom" || from === "fromTop" ? 'h-screen' : 'h-full'} ${trackWidth ? `w-[${trackWidth}]` : from === "fromLeft" || from === "fromRight" ? 'w-screen' : 'w-full'}`}>
        {children}
      </div>
    </div>
  )
};