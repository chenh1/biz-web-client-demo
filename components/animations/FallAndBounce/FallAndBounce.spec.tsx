import { FallAndBounce } from "./FallAndBounce";
import { render, waitFor } from '@testing-library/react';
import { JSDOM } from "jsdom"
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window

const observe = jest.fn();
const unobserve = jest.fn();

// you can also pass the mock implementation
// to jest.fn as an argument
global.IntersectionObserver = jest.fn(function (callback, options) {
  return {
    observe,
    unobserve,
    root: null,
    rootMargin: '',
    thresholds: [],
    disconnect: jest.fn(),
    takeRecords: jest.fn(),
  };
});


it('falls and bounces when in view', async () => {
  const { getByTestId, asFragment } = render(
    <FallAndBounce from="fromLeft" disable={false}>
      <p>FallAndBounce</p>
    </FallAndBounce>
  );

  await waitFor(() => {
    expect(asFragment()).toMatchSnapshot();
  });
});