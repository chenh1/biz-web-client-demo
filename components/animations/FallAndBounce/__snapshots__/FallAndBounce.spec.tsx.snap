// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`falls and bounces when in view 1`] = `
<DocumentFragment>
  <div
    class="relative  "
  >
    <div
      class="fallAndBounce freeze absolute fromLeft right-0 h-full w-screen"
    >
      <p>
        FallAndBounce
      </p>
    </div>
  </div>
</DocumentFragment>
`;
